Fast Reload Option

wcawijngaards opened the pull request:
The branch continues the fast-reload work, and adds more config options to fast-reload. This is a good idea, since otherwise, old references cause trouble, that would have to be resolved in another manner if more config options are not supported by reloading them.

wcawijngaards committed:
- fast-reload, add unbound-control fast_reload

wcawijngaards committed:
- fast-reload, make a thread to service the unbound-control command.

wcawijngaards committed:
- fast-reload, communication sockets for information transfer.

wcawijngaards committed:
- fast-reload, fix compile for unbound-dnstap-socket.

wcawijngaards committed:
- fast-reload, set nonblocking communication to keep the server thread
  responding to DNS requests.

wcawijngaards committed:
- fast-reload, poll routine to test for readiness, timeout fails connection.

wcawijngaards committed:
- fast-reload, detect loop in sock_poll_timeout routine.

wcawijngaards committed:
- fast-reload, send done and exited notification.

wcawijngaards committed:
- fast-reload, defines for constants in ipc.

wcawijngaards committed:
Merge branch 'master' into fast-reload

wcawijngaards committed:
- fast-reload, ipc socket recv and send resists partial reads and writes and
  can continue byte by byte. Also it can continue after an interrupt.

wcawijngaards committed:
- fast-reload, send exit command to thread when done.

wcawijngaards committed:
- fast-reload, output strings for client on string list.

wcawijngaards committed:
- fast-reload, add newline to terminal output.

wcawijngaards committed:
- fast-reload, send client string to remote client.

wcawijngaards committed:
- fast-reload, better debug output.

wcawijngaards committed:
- fast-reload, print queue structure, for output to the remote client.

wcawijngaards committed:
Merge branch 'master' into fast-reload

wcawijngaards committed:
- fast-reload, move print items to print queue from fast_reload_thread struct.

wcawijngaards committed:
- fast-reload, keep list of pending print queue items in daemon struct.

wcawijngaards committed:
- fast-reload, comment explains in_list for printq to print remainder.

wcawijngaards committed:
- fast-reload, unit test testdata/fast_reload_thread.tdir that tests the
  thread output.

wcawijngaards committed:
Merge branch 'master' into fast-reload

wcawijngaards committed:
- fast-reload, fix test link for fast_reload_printq_list_delete function.

wcawijngaards committed:
- fast-reload, reread config file from disk.

wcawijngaards committed:
- fast-reload, unshare forwards, making the structure locked, with an rwlock.

wcawijngaards committed:
- fast-reload, for nonthreaded, the unbound-control commands forward,
  forward_add and forward_delete should be distributed to other processes,
  but when threaded, they should not be distributed to other threads because
  the structure is not thread specific any more.

wcawijngaards committed:
Merge branch 'master' into fast-reload

wcawijngaards committed:
- fast-reload, unshared stub hints, making the structure locked, with an rwlock.

wcawijngaards committed:
Merge branch 'master' into fast-reload

wcawijngaards committed:
- fast-reload, helpful comments for hints lookup function return value.

wcawijngaards committed:
- fast-reload, fix bug in fast reload printout, the strlist appendlist routine,
  and printout time statistics after the reload is done.

wcawijngaards committed:
- fast-reload, keep track of reloadtime and deletestime and print them.

wcawijngaards committed:
- fast-reload, keep track of constructtime and print it.

wcawijngaards committed:
- fast-reload, construct new items.

wcawijngaards committed:
- fast-reload, better comment.

wcawijngaards committed:
- fast-reload, reload the config and swap trees for forwards and stub hints.

wcawijngaards committed:
- fast-reload, in forwards_swap_tree set protection of trees with locks.

wcawijngaards committed:
- fast-reload, in hints_swap_tree also swap the node count of the trees.

wcawijngaards committed:
- fast-reload, reload ipc to stop and start threads.

wcawijngaards committed:
Merge branch 'master' into fast-reload

wcawijngaards committed:
- fast-reload, unused forward declarations removed.

wcawijngaards committed:
- fast-reload, unit test that fast reload works with forwards and stubs.

wcawijngaards committed:
Merge branch 'master' into fast-reload

wcawijngaards committed:
Merge branch 'master' into fast-reload

wcawijngaards committed:
- fast-reload, fix clang analyzer warnings.

wcawijngaards committed:
Merge branch 'master' into fast-reload

wcawijngaards committed:
Merge branch 'master' into fast-reload

wcawijngaards committed:
- fast-reload, small documentation entry in unbound-control -h output.

wcawijngaards committed:
- fast-reload, printout memory use by fast reload, in bytes.

wcawijngaards committed:
Merge branch 'master' into fast-reload

wcawijngaards committed:
- fast-reload, compile without threads.

wcawijngaards committed:
- fast-reload, document fast_reload in man page.

wcawijngaards committed:
- fast-reload, print ok when done successfully.

wcawijngaards committed:
- fast-reload, option for fast-reload commandline, +v verbosity option,
  with timing and memory use output.

wcawijngaards committed:
- fast-reload, option for fast-reload commandline, +p does not pause threads.

wcawijngaards committed:
- fast-reload, option for fast-reload commandline, +d drops mesh queries.

wcawijngaards committed:
- fast-reload, fix to poll every thread with nopause to make certain that
  resources are not held by the threads and can be deleted.

wcawijngaards committed:
- fast-reload, fix to use atomic store for config variables with nopause.

wcawijngaards committed:
- fast-reload, reload views.

wcawijngaards committed:
Merge branch 'master' into fast-reload

wcawijngaards committed:
- fast-reload, when tag defines are different, it drops the queries.

wcawijngaards committed:
- fast-reload, fix tag define check.

wcawijngaards committed:
Merge branch 'master' into fast-reload

wcawijngaards committed:
- fast-reload, document that tag change causes drop of queries.

wcawijngaards committed:
- fast-reload, fix space in documentation man page.

wcawijngaards committed:
Merge branch 'master' into fast-reload-option

wcawijngaards committed:
- fast-reload, copy respip client information to query state, put views tree
  in module env for lookup.

wcawijngaards committed:
- fast-reload, nicer respip view comparison.

wcawijngaards committed:
- fast-reload, respip global set is in module env.

wcawijngaards committed:
- fast-reload, document that respip_client_info acl info is copied.

wcawijngaards committed:
- fast-reload, reload the respip_set.

wcawijngaards committed:
- fast-reload, document no pause and pick up of use_response_ip boolean.

wcawijngaards committed:
- fast-reload, fix test compile.

gthess committed:
Update locking management for iter_fwd and iter_hints methods. (#1054)

fast reload, move most of the locking management to iter_fwd and
iter_hints methods. The caller still has the ability to handle its
own locking, if desired, for atomic operations on sets of different
structs.

Co-authored-by: Wouter Wijngaards <wcawijngaards@users.noreply.github.com>

wcawijngaards committed:
- fast-reload, reload local zones.

wcawijngaards committed:
Merge branch 'master' into fast-reload-option

wcawijngaards committed:
- fast-reload, reload access-control.

wcawijngaards committed:
- fast-reload, reload access control interface, such as interface-action.

wcawijngaards committed:
- fast-reload, reload tcp-connection-limit.

wcawijngaards committed:
- fast-reload, improve comments on acl_list and tcl_list swap tree.

wcawijngaards committed:
- fast-reload, fixup references to old tcp connection limits in open tcp
  connections.

wcawijngaards committed:
- fast-reload, fixup to clean tcp connection also for different linked order.

wcawijngaards committed:
- fast-reload, if no tcp connection limits existed, no need to remove
  references for that.

wcawijngaards committed:
- fast-reload, document more options that work and do not work.

wcawijngaards committed:
- fast-reload, reload auth_zone and rpz data.

wcawijngaards committed:
Merge branch 'master' into fast-reload-option

wcawijngaards committed:
- fast-reload, fix auth_zones_get_mem.

wcawijngaards committed:
- fast-reload, fix compilation of testbound for the new comm_timer_get_mem
  reference in remote control.

wcawijngaards committed:
- fast-reload, change use_rpz with reload.

wcawijngaards committed:
Merge branch 'master' into fast-reload-option

wcawijngaards committed:
- fast-reload, list changes in auth zones and stop zonemd callbacks for
  deleted auth zones.

wcawijngaards committed:
- fast-reload, note xtree is not swapped, and why it is not swapped.

wcawijngaards committed:
Merge branch 'master' into fast-reload-option

wcawijngaards committed:
- fast-reload, for added auth zones, pick up zone transfer and zonemd tasks.

wcawijngaards committed:
- fast-reload, unlock xfr when done with transfer pick up.

wcawijngaards committed:
- fast-reload, unlock z when picking up the xfr for it during transfer task
  pick up.

wcawijngaards committed:
- fast-reload, pick up task changes for added, deleted and modified auth zones.

wcawijngaards committed:
Merge branch 'master' into fast-reload-option

wcawijngaards committed:
- fast-reload, remove xfr of auth zone deletion without tasks.

wcawijngaards committed:
Merge branch 'master' into fast-reload-option

wcawijngaards committed:
- fast-reload, pick up zone transfer config.

wcawijngaards committed:
Merge branch 'master' into fast-reload-option

wcawijngaards committed:
- fast-reload, the main worker thread picks up the transfer tasks and also
  performs setup of the xfer struct.

wcawijngaards committed:
- fast-reload, keep writelock on newzone when auth zone changes.

rcmcdonald91 commented:
Any updates on this?

wcawijngaards commented:
If you are eager to use, the branch is somewhat in a working state. At certain commits it works, passes tests, and reloads a growing list of configuration options.

wcawijngaards committed:
- fast-reload, change cachedb_enabled setting.

wcawijngaards committed:
Merge branch 'master' into fast-reload-option

wcawijngaards committed:
- fast-reload, pick up edns-strings config.

wcawijngaards committed:
- fast-reload, note that settings are not updated.

wcawijngaards committed:
- fast-reload, pick up dnstap config.

wcawijngaards committed:
Merge branch 'master' into fast-reload-option

wcawijngaards committed:
- fast-reload, dnstap options that need to be loaded without +p.

wcawijngaards committed:
- fast-reload, fix auth zone reload

wcawijngaards committed:
- fast-reload, remove debug for auth zone test.

wcawijngaards committed:
- fast-reload, fix auth zone reload with zone transfer.

wcawijngaards committed:
- fast-reload, fix auth zone reload lock order.

wcawijngaards committed:
- fast-reload, remove debug from fast reload test.

wcawijngaards committed:
- fast-reload, remove unused function.

wcawijngaards committed:
Merge branch 'master' into fast-reload-option

wcawijngaards committed:
- fast-reload, fix the worker trust anchor probe timer lock acquisition in
  the probe answer callback routine for trust anchor probes.

wcawijngaards committed:
- fast-reload, reload trust anchors.

wcawijngaards committed:
- fast-reload, fix trust anchor reload lock on autr global data and test
  for trust anchor reload.

wcawijngaards committed:
- fast-reload, adjust cache sizes.

wcawijngaards committed:
- fast-reload, reload cache sizes when changed.

wcawijngaards committed:
Merge branch 'master' into fast-reload-option

wcawijngaards committed:
- fast-reload, reload validator env changes.

wcawijngaards committed:
- fast-reload, reload mesh changes.

wcawijngaards committed:
Merge branch 'master' into fast-reload-option

wcawijngaards committed:
- fast-reload, check for incompatible changes.

wcawijngaards committed:
- fast-reload, improve error text for incompatible change.

wcawijngaards committed:
- fast-reload, fix check config option compatibility.

wcawijngaards committed:
- fast-reload, improve error text for nopause change.

wcawijngaards committed:
- fast-reload, fix spelling of incompatible options.

wcawijngaards committed:
- fast-reload, reload target-fetch-policy, outbound-msg-retry, max-sent-count
  and max-query-restarts.

wcawijngaards committed:
Merge branch 'master' into fast-reload-option

wcawijngaards committed:
- fast-reload, check nopause config change for target-fetch-policy.

wcawijngaards committed:
- fast-reload, reload do-not-query-address, private-address and capt-exempt.

wcawijngaards committed:
- fast-reload, check nopause config change for do-not-query-address,
  private-address and capt-exempt.

wcawijngaards committed:
- fast-reload, check fast reload not possible due to interface and
  outgoing-interface changes.

wcawijngaards committed:
- fast-reload, reload nat64 settings.

wcawijngaards committed:
- fast-reload, reload settings stored in the infra structure.

wcawijngaards committed:
- fast-reload, fix modstack lookup and remove outgoing-range check.

wcawijngaards committed:
- fast-reload, more explanation for config parse failure.

wcawijngaards committed:
Merge branch 'master' into fast-reload-option

wcawijngaards committed:
- fast-reload, reload worker outside network changes.

wcawijngaards committed:
- fast-reload, detect incompatible changes in network settings.

wcawijngaards committed:
fast-reload, commit test files.

wcawijngaards committed:
Merge branch 'master' into fast-reload-option

wcawijngaards committed:
Merge branch 'master' into fast-reload-option

wcawijngaards committed:
- fast-reload, fix warnings for call types in windows compile.

wcawijngaards committed:
- fast-reload, fix warnings and comm_point_internal for tcp wouldblock calls.

wcawijngaards committed:
Merge branch 'master' into fast-reload-option

wcawijngaards committed:
Merge branch 'master' into fast-reload-option

wcawijngaards committed:
- fast-reload, extend lock checks for repeat thread ids.

wcawijngaards committed:
- fast-reload, additional test cases, cache change and tag changes.

wcawijngaards committed:
Merge branch 'master' into fast-reload-option

wcawijngaards committed:
- fast-reload, fix documentation for auth_zone_verify_zonemd_with_key.

wcawijngaards committed:
- fast-reload, fix copy_cfg type casts and memory leak on config parse failure.

wcawijngaards committed:
Merge branch 'master' into fast-reload-option

wcawijngaards committed:
- fast-reload, fix use of WSAPoll.

gthess committed:
Review comments for the fast reload feature (#1259)

* - fast-reload review, respip set can be null from a view.

* - fast-reload review, typos.

* - fast-reload review, keep clang static analyzer happy.

* - fast-reload review, don't forget to copy tag_actions.

* - fast-reload review, less indentation.

* - fast-reload review, don't leak respip_actions when reloading.

* - fast-reload review, protect NULL pointer dereference in get_mem
  functions.

* - fast-reload review, add fast_reload_most_options.tdir to test most
  options with high verbosity when fast reloading.

* - fast-reload review, don't skip new line on long error printouts.

* - fast-reload review, typo.

* - fast-reload review, use new_z for consistency.

* - fast-reload review, nit for unlock ordering to make eye comparison
  with the lock counterpart easier.

* - fast-reload review, in case of error the sockets are already closed.

* - fast-reload review, identation.

* - fast-reload review, add static keywords.

* - fast-reload review, update unbound-control usage text.

* - fast-reload review, updates to the man page.

gthess reviewed (approved):
Looks good to me!
My review comments were part of https://github.com/NLnetLabs/unbound/pull/1259 which are already addressed here.
When conflicts are resolved, a squash commit should land this feature into master and onto the next release :+1:

wcawijngaards committed:
- fast-reload, the fast-reload command is experimental.

wcawijngaards committed:
Merge branch 'master' into fast-reload-option

wcawijngaards committed:
- fast-reload, fix compile of doqclient for fast reload functions.

wcawijngaards committed:
Changelog comment for #1042
- Merge #1042: Fast Reload. The unbound-control fast_reload is added.
  It reads changed config in a thread, then only briefly pauses the
  service threads, that keep running. DNS service is only interrupted
  briefly, less than a second.
